src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -450,6 +450,146 @@ struct TreeTraversal<BVH, Predicates, Callback, NearestPredicateTag>
+          *this);
+    }
+  }
+
+  struct OneLeafTree
+  {};
+
+  KOKKOS_FUNCTION void operator()(OneLeafTree, int queryIndex) const
+  {
+    auto const &predicate = Access::get(_predicates, queryIndex);
+    auto const root = HappyTreeFriends::getRoot(_bvh);
+    auto const &root_bounding_volume =
+        HappyTreeFriends::getBoundingVolume(_bvh, root);
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+    if (distance(getGeometry(predicate), root_bounding_volume) != inf)
+    {
+      _callback(predicate, 0);
+    }
+  }
+
+  KOKKOS_FUNCTION void operator()(int queryIndex) const
+  {
+    auto const &predicate = Access::get(_predicates, queryIndex);
+    using ArborX::Details::HappyTreeFriends;
+
+    auto const distance = [geometry = getGeometry(predicate),
+                           bvh = _bvh](int node) {
+      auto const &box = HappyTreeFriends::getBoundingVolume(bvh, node);
+      using Details::distance;
+      return distance(geometry, box);
+    };
+
+    using PairIndexDistance = Kokkos::pair<int, float>;
+    struct CompareDistance
+    {
+      KOKKOS_FUNCTION bool operator()(PairIndexDistance const &lhs,
+                                      PairIndexDistance const &rhs) const
+      {
+        return lhs.second > rhs.second;
+      }
+    };
+
+    PairIndexDistance heap[64];
+    PairIndexDistance *heap_last = heap;
+    CompareDistance const compare;
+    constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+
+    int node = HappyTreeFriends::getRoot(_bvh);
+    int left_child;
+    int right_child;
+
+    while (true)
+    {
+      if (HappyTreeFriends::isLeaf(_bvh, node))
+      {
+        if (invoke_callback_and_check_early_exit(
+                _callback, predicate,
+                HappyTreeFriends::getLeafPermutationIndex(_bvh, node)))
+          return;
+
+        if (heap == heap_last)
+          break; // heap is empty
+
+        node = heap->first;
+        popHeap(heap, heap_last--, compare);
+      }
+      else
+      {
+        left_child = HappyTreeFriends::getLeftChild(_bvh, node);
+        right_child = HappyTreeFriends::getRightChild(_bvh, node);
+
+        float const distance_left = distance(left_child);
+        auto const left_pair = Kokkos::make_pair(left_child, distance_left);
+
+        float const distance_right = distance(right_child);
+        auto const right_pair = Kokkos::make_pair(right_child, distance_right);
+
+        auto const &closer_pair =
+            distance_left < distance_right ? left_pair : right_pair;
+        auto const &further_pair =
+            distance_left < distance_right ? right_pair : left_pair;
+
+        if (heap != heap_last && heap->second < closer_pair.second)
+        {
+          node = heap->first;
+          popHeap(heap, heap_last--, compare);
+          if (closer_pair.second < inf)
+          {
+            *heap_last++ = closer_pair;
+            pushHeap(heap, heap_last, compare);
+          }
```

Comment:
This seems off. You have the same code below (583-586), so it seems that the further pair may be pushed twice into the heap. I think it may be fine to keep it here, and try to use `popPush` instead of separate `popHeap` and `pushHeap`. But then the other instance would have be fully in the `else` branch.

Reply:
The if-clause from 571 to 582 is solely concerned with the closer pair while 583 to 586 is responsible for the further pair. I don't see how the further pair could be pushed twice.

Reply:
Ah, my bad. I misread the code, thinking that the `if` clauses were the same. Then, just try to use `popPush`.

Reply:
Just introducing another branch here makes the benchmark here 25% slower and using `popPush` doesn't seem to change anything after that.